knotx-core/src/main/java/io/knotx/splitter/HtmlFragmentSplitter.java

```diff
@@ -51,19 +60,62 @@
         idx = matchResult.end();
       }
       if (idx < html.length()) {
-        fragments.add(toRaw(html, idx, html.length()));
+        processMarkup(fragments, fallbackMarkers, html, idx, html.length());
       }
     } else {
-      fragments.add(toRaw(html, 0, html.length()));
+      processMarkup(fragments, fallbackMarkers, html, 0, html.length());
     }
     return fragments;
   }
 
-  private Fragment toRaw(String html, int startIdx, int endIdx) {
-    return Fragment.raw(html.substring(startIdx, endIdx));
+  private void processMarkup(List<Fragment> fragments, List<FallbackMarker> fallbackMarkers, String html, int startIdx, int endIdx) {
+    int idx = startIdx;
+    for (FallbackMarker fe : fallbackMarkers) {
+      if (idx > fe.start) {
+        continue;
+      } else if (fe.start > endIdx) {
+        break;
+      } else if (idx < fe.start) {
+        fragments.add(Fragment.raw(html.substring(idx, fe.start)));
+      }
+      fragments.add(toFallback(html, fe.start, fe.end));
```

Comment:
Maybe it would be worth to keep additional field in the `Fragment` e.g. `fragmentId` and extract it here?
That will allow to not use markup processing (JSoup) at the Assembler level to extract fallback id.

Reply:
I like this idea
`data-knotx-fragment-id` - should hold id of current fragment (new parameter)
`data-knotx-fallback` - should hold id of a other (fallback) fragment

Reply:
I'm having some second thoughts on this, it would require additional capture groups in the regexps. I'm afraid we would take a performance hit here.

Reply:
Finally I decided to remove JSoup parsing from the Assembler. 
Additional regex to extract the fragment id is run only against fallback snippets.